orb-ui/src/engine/animations/simple_spinner.rs

```diff
@@ -0,0 +1,218 @@
+use std::f64::consts::PI;
+
+const SPIN_SPEED_SECONDS_PER_TURN: f64 = 16.0;
+
+/// Animated spinner.
+#[derive(Clone)]
+pub struct SimpleSpinner<const N: usize> {
+    // radians
+    phase: f64,
+    color: Argb,
+    background: Argb,
+    // rad/sec
+    speed: f64,
+    transition: Option<Transition>,
+    transition_time: f64,
+    transition_background: Option<Argb>,
+}
+
+impl<const N: usize> SimpleSpinner<N> {
+    /// Creates a new [`SimpleSpinner`] with one arc.
+    #[allow(dead_code)]
+    #[must_use]
+    pub fn new(color: Argb, background: Option<Argb>) -> Self {
+        Self {
+            speed: 2.0 * PI / SPIN_SPEED_SECONDS_PER_TURN,
+            phase: PI / 2.0, // start animation at 12 o'clock
+            color,
+            background: background.unwrap_or(Argb::OFF),
+            transition: None,
+            transition_time: 0.0,
+            transition_background: None,
+        }
+    }
+
+    /// Set the speed of the spinner in radians per second.
+    #[allow(dead_code)]
+    pub fn speed(self, speed: f64) -> Self {
+        Self { speed, ..self }
+    }
+
+    /// Get the current phase of the spinner.
+    /// The phase is the current position of the spinner in radians.
+    pub fn phase(&self) -> f64 {
+        self.phase
+    }
+
+    #[allow(dead_code)]
+    pub fn fade_in(self, duration: f64) -> Self {
+        Self {
+            transition: Some(Transition::FadeIn(duration)),
+            transition_time: 0.0,
+            ..self
+        }
+    }
+
+    pub fn background(&self) -> Argb {
+        self.background
+    }
+}
+
+impl<const N: usize> Animation for SimpleSpinner<N> {
+    type Frame = RingFrame<N>;
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    #[allow(clippy::cast_precision_loss, clippy::float_cmp)]
+    fn animate(
+        &mut self,
+        frame: &mut RingFrame<N>,
+        dt: f64,
+        idle: bool,
+    ) -> AnimationState {
+        if let Some(Transition::ForceStop) = self.transition {
+            return AnimationState::Finished;
+        }
+
+        if let Some(Transition::StartDelay(duration)) = self.transition {
+            self.transition_time += dt;
+            if self.transition_time >= duration {
+                self.transition = None;
+            }
+            return AnimationState::Running;
+        }
+
+        self.phase = (self.phase + dt * self.speed) % (2.0 * PI);
+        // `N - led-index` with `led-index` increasing in {0,N}, because the LED strip goes anti-clockwise
+        // `N as f64 * 3.0 / 4.0` because the first LED is at 6 o'clock (3PI/2)
+        let progress = (N as f64 - (self.phase * N as f64 / (2.0 * PI))
+            + N as f64 * 3.0 / 4.0)
+            % N as f64;
+        let led_index = progress as usize;
+        let head_tail_scale = progress - led_index as f64;
+
+        if !idle {
+            let mut factor = match self.transition {
```

Comment:
factor for what? consider naming more descriptively. Perhaps "tween_factor"

Reply:
renamed to `scaling_factor`